xonsh/commands_cache.py

```diff
@@ -0,0 +1,142 @@
+        # did aliases change?
+        alss = getattr(builtins, 'aliases', set())
+        al_hash = hash(frozenset(alss))
+        cache_valid = cache_valid and al_hash == self._alias_checksum
+        self._alias_checksum = al_hash
+        # did the contents of any directory in PATH change?
+        max_mtime = 0
+        for path in pathset:
+            mtime = os.stat(path).st_mtime
+            if mtime > max_mtime:
+                max_mtime = mtime
+        cache_valid = cache_valid and (max_mtime <= self._path_mtime)
+        self._path_mtime = max_mtime
+        if cache_valid:
+            return self._cmds_cache
+        allcmds = {}
+        for path in reversed(paths):
+            # iterate backwards so that entries at the front of PATH overwrite
+            # entries at the back.
+            for cmd in executables_in(path):
+                key = cmd.upper() if ON_WINDOWS else cmd
+                allcmds[key] = (os.path.join(path, cmd), cmd in alss)
+        only_alias = (None, True)
+        for cmd in alss:
+            if cmd not in allcmds:
+                allcmds[cmd] = only_alias
+        self._cmds_cache = allcmds
+        return allcmds
+
+    def lazyin(self, key):
+        """Checks if the value is in the current cache without the potential to
+        update the cache. It just says whether the value is known *now*. This
+        may not reflect precisely what is on the $PATH.
+        """
+        return key in self._cmds_cache
+
+    def lazyiter(self):
+        """Returns an iterator over the current cache contents without the
+        potential to update the cache. This may not reflect what is on the
+        $PATH.
+        """
+        return iter(self._cmds_cache)
+
+    def lazylen(self):
+        """Returns the length of the current cache contents without the
+        potential to update the cache. This may not reflect precisely
+        what is on the $PATH.
+        """
+        return len(self._cmds_cache)
+
+    def lazyget(self, key, default=None):
+        """A lazy value getter."""
+        return self._cmds_cache.get(key, default)
+
+    def locate_binary(self, name):
+        """Locates an executable on the file system using the cache."""
+        # make sure the cache is up to date by accessing the property
+        _ = self.all_commands
+        return self.lazy_locate_binary(name)
+
+    def lazy_locate_binary(self, name):
+        """Locates an executable in the cache, without checking its validity."""
+        possibilities = self.get_possible_names(name)
+
+        if ON_WINDOWS:
+            # Windows users expect to be able to execute files in the same
+            # directory without `./`
+            cwd = _get_cwd()
+            local_bin = next((
+                full_name for full_name in possibilities
+                if os.path.isfile(full_name)
+            ), None)
+            if local_bin:
+                return os.path.abspath(os.path.relpath(local_bin, cwd))
```

Comment:
I know this is already merged. But I think could be simplified a bit. Especially if @scopatz want to leave out the `_get_cwd()` bit. `os.path.abspath()` already works relative to current directory. So we could just do:
```python 
return os.path.abspath(local_bin)
```